Exception handling for generateFeatures

kathrynkodama opened the pull request:
Handles exception for the `generateFeatures` task.

When the task runs standalone and an error occurs, a `GradleException` is thrown and users will see a build failure.
```
kathrynkodama@MacBook-Pro demo-devmode-scan % gradle generateFeatures

> Task :generateFeatures FAILED
Parsing the server file /Users/kathrynkodama/devex/featureGeneration/demo-devmode-scan/src/main/liberty/config/server.xml1 for features and includes.
SEVERE: CWMIG12083E: The cdi-2.0 feature requires the io.openliberty.servlet.api-4.0 feature which conflicts with the io.openliberty.servlet.api-3.1 feature that is required by the batch-1.0 feature.
SEVERE: CWMIG12083E: The cdi-2.0 feature requires the io.openliberty.servlet.api-4.0 feature which conflicts with the io.openliberty.servlet.api-3.1 feature that is required by the batch-1.0 feature.
SEVERE: CWMIG12083E: The cdi-2.0 feature requires the io.openliberty.servlet.api-4.0 feature which conflicts with the io.openliberty.servlet.api-3.1 feature that is required by the batch-1.0 feature.
SEVERE: CWMIG12085E: The feature list for the  application cannot be generated because of the conflicts between required features.
SEVERE: CWMIG12085E: The feature list for the  application cannot be generated because of the conflicts between required features.
SEVERE: CWMIG12085E: The feature list for the  application cannot be generated because of the conflicts between required features.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateFeatures'.
> A working set of features could not be generated due to conflicts between configured features and the application's API usage: [cdi-2.0, batch-1.0, webProfile-8.0, jpa-2.2, jaxrs-2.1]. Review and update your server configuration and application to ensure they are not using conflicting features and APIs from different levels of MicroProfile, Java EE, or Jakarta EE. Refer to the following set of suggested features for guidance: [cdi-2.0, mpHealth-2.2, batch-1.0, jpa-2.2, jaxrs-2.1]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
1 actionable task: 1 executed
```

When running the goal as part of dev mode:

On startup if `generateFeatures` throws an error dev mode will exit. If there is a feature conflict (most likely reason generate-features will fail) we fail here first to show users a more relevant error message rather than continuing on to `installFeature` and failing there. Error message indicates that users can disable feature generation with --generateFeatures=false

```
> Task :generateFeatures FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateFeatures'.
> A working set of features could not be generated due to conflicts between configured features and the application's API usage: [cdi-2.0, batch-1.0, webProfile-8.0, jpa-2.2, jaxrs-2.1]. Review and update your server configuration and application to ensure they are not using conflicting features and APIs from different levels of MicroProfile, Java EE, or Jakarta EE. Refer to the following set of suggested features for guidance: [cdi-2.0, mpHealth-2.2, batch-1.0, jpa-2.2, jaxrs-2.1]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 2s
1 actionable task: 1 executed

> Task :libertyDev FAILED
To disable the automatic generation of features, start dev mode with --generateFeatures=false.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libertyDev'.
> Could not execute build using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-6.9.1-bin.zip'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 9s
1 actionable task: 1 executed
```

Once in the inner dev mode loop errors from `generateFeatures` will not fail dev mode, users will see the following error in their console:
```
> Task :generateFeatures FAILED
Parsing the server file /Users/kathrynkodama/devex/featureGeneration/demo-devmode-scan/src/main/liberty/config/server.xml1 for features and includes.
Parsing the server file /Users/kathrynkodama/devex/featureGeneration/demo-devmode-scan/src/main/liberty/config/configDropins/overrides/generated-features.xml for features and includes.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateFeatures'.
> A working set of features could not be generated due to conflicts between configured features: [cdi-2.0, mpHealth-2.2, jpa-2.2, webProfile-7.0, jaxrs-2.1]. Review and update your server configuration to ensure it is not using conflicting features from different levels of MicroProfile, Java EE, or Jakarta EE. Refer to the following set of suggested features for guidance: [cdi-2.0, mpHealth-2.2, batch-1.0, jpa-2.2, jaxrs-2.1]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 879ms
1 actionable task: 1 executed

> Task :libertyDev
Could not execute build using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-6.9.1-bin.zip'..
 To disable the automatic generation of features, type 'g' and press Enter.
```

Signed-off-by: Kathryn Kodama <kathryn.s.kodama@gmail.com>

TrevCraw reviewed (commented):
Similar to ci.maven, the message `To disable the automatic generation of features, start dev mode with --generateFeatures=false.` is separated from the actual error and may give users the wrong impression. Is it possible to put that message next to the error message?

kathrynkodama commented:
Updated the error message shown when `generateFeatures` fails on dev mode startup to:
```
> Task :generateFeatures FAILED
Parsing the server file /Users/kathrynkodama/devex/featureGeneration/demo-devmode-scan/src/main/liberty/config/server.xml1 for features and includes.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateFeatures'.
> A working set of features could not be generated due to conflicts between configured features and the 
application's API usage: [cdi-2.0, batch-1.0, webProfile-8.0, jpa-2.2, jaxrs-2.1]. Review and update your server 
configuration and application to ensure they are not using conflicting features and APIs from different levels of 
MicroProfile, Java EE, or Jakarta EE. Refer to the following set of suggested features for guidance: [cdi-2.0, 
mpHealth-2.2, batch-1.0, jpa-2.2, jaxrs-2.1]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 2s
1 actionable task: 1 executed
> Task :libertyDev FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libertyDev'.
> Execution failed for task ':generateFeatures'. To disable the automatic generation of features, start dev mode 
with --generateFeatures=false.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 8s
1 actionable task: 1 executed
```

Also updated the comments inserted in the `generated-features.xml` file. See https://github.com/OpenLiberty/ci.maven/pull/1331#issuecomment-991780009
